### ISSUE TITLE
git_diff_foreach should be a superset of git_diff_print functionality

### ISSUE THREAD

**Issue by eaigner** [libgit2/libgit2#5341]
I currently have a case where I manually create a patch using `git_diff_foreach` for selective line staging. Now to apply the patch with `git`, I also need to generate a patch header from the `diff`.

I currently solved this by separately emitting the patch header using `git_diff_print` and prepend it, which I don't really like, since it seems kind of fragile.

I could reproduce the logic in `git_diff_delta__format_file_header` to manually create the header myself from a `git_diff_delta`, but I'd be obviously better if this could be entirely handled within `libgit2`.

So my concrete problem is the following:

- `git_diff_foreach` emits `git_diff_binary` but no patch header lines.
- `git_diff_print` emits patch header lines, but no `git_diff_binary` structs.

I think the simplest solution would just be to emit the `git_diff_binary` object with the single print callback.

**Comment by pks-t** [comment]
> git_diff_foreach emits git_diff_binary but no patch header lines.

I agree it does make sense to expose such functionality. It does seem very specific, though. What about a more generic function that, given a diff delta, will format it accordingly? Ideally, it should be get an options struct that allows one to specify various options so that one can e.g. print only the file header or only the diff contents.

> git_diff_print emits patch header lines, but no git_diff_binary structs.

Feels like `git_diff_print` should have an option to print binary diffs, too.
